File: impl/go/internal/engine/engine_test.go
package engine

import (
	"testing"

	"github.com/zaibatsu/zaibatsu-go/internal/data"
	"github.com/zaibatsu/zaibatsu-go/internal/domain"
)

func loadOrSkip(t *testing.T) *domain.GameData {
	t.Helper()
	gd, err := data.LoadDefault("speedrunners")
	if err != nil {
		t.Fatalf("load speedrunners data: %v", err)
	}
	return gd
}

func TestControlMarkersFor(t *testing.T) {
	gd := loadOrSkip(t)
	cases := map[int]int{2: 10, 3: 8, 4: 6}
	for n, want := range cases {
		if got := ControlMarkersFor(gd.Mode, n); got != want {
			t.Errorf("ControlMarkersFor(%d) = %d, want %d", n, got, want)
		}
	}
	// Unlisted count falls back to the mode default.
	if got := ControlMarkersFor(gd.Mode, 7); got != 8 {
		t.Errorf("ControlMarkersFor(7) fallback = %d, want 8 (default)", got)
	}
}

func TestNewGameSetup(t *testing.T) {
	gd := loadOrSkip(t)
	s, err := NewGame(Config{Data: gd, PlayerNames: []string{"Alice", "Bob"}, Seed: 42})
	if err != nil {
		t.Fatalf("NewGame: %v", err)
	}
	if len(s.Players) != 2 {
		t.Fatalf("expected 2 players, got %d", len(s.Players))
	}
	for _, p := range s.Players {
		if p.ControlMarkersTotal != 10 {
			t.Errorf("player %s markers = %d, want 10", p.ID, p.ControlMarkersTotal)
		}
		if p.PawnID == "" {
			t.Errorf("player %s has no pawn assigned", p.ID)
		}
	}
	// Distinct starter pawns.
	if s.Players[0].PawnID == s.Players[1].PawnID {
		t.Errorf("players got the same starter pawn %q", s.Players[0].PawnID)
	}
	// Opening deal is asymmetric: p1 gets fewer than p2.
	if len(s.Players[0].Hand) >= len(s.Players[1].Hand) {
		t.Errorf("expected asymmetric opening deal, got p1=%d p2=%d",
			len(s.Players[0].Hand), len(s.Players[1].Hand))
	}
}

func TestDeterminism(t *testing.T) {
	gd := loadOrSkip(t)
	mk := func() *domain.GameState {
		s, _ := NewGame(Config{Data: gd, PlayerNames: []string{"A", "B", "C"}, Seed: 7})
		return s
	}
	a, b := mk(), mk()
	for i := range a.Players {
		if a.Players[i].PawnID != b.Players[i].PawnID {
			t.Fatalf("same seed produced different pawn assignment at %d", i)
		}
	}
	if len(a.Deck) == 0 || a.Deck[0] != b.Deck[0] {
		t.Fatalf("same seed produced different deck order")
	}
}

func TestRecycleRefillsHand(t *testing.T) {
	gd := loadOrSkip(t)
	s, _ := NewGame(Config{Data: gd, PlayerNames: []string{"A", "B"}, Seed: 1})
	// Player 1 opened with fewer than max; after their turn's recycle they should be at max.
	if err := RunTurn(s, gd, []Action{{Type: ActPass}}); err != nil {
		t.Fatalf("RunTurn: %v", err)
	}
	// After RunTurn, current player advanced to p2; inspect p1 (index 0).
	if got := len(s.Players[0].Hand); got != s.Players[0].MaxHandSize {
		t.Errorf("p1 hand after recycle = %d, want %d", got, s.Players[0].MaxHandSize)
	}
}

func TestTurnAdvances(t *testing.T) {
	gd := loadOrSkip(t)
	s, _ := NewGame(Config{Data: gd, PlayerNames: []string{"A", "B", "C"}, Seed: 3})
	if s.CurrentPlayer != 0 {
		t.Fatalf("expected to start on player 0")
	}
	_ = RunTurn(s, gd, nil)
	if s.CurrentPlayer != 1 {
		t.Errorf("after one turn CurrentPlayer = %d, want 1", s.CurrentPlayer)
	}
	_ = RunTurn(s, gd, nil)
	_ = RunTurn(s, gd, nil)
	if s.CurrentPlayer != 0 {
		t.Errorf("after three turns CurrentPlayer = %d, want 0 (wrapped)", s.CurrentPlayer)
	}
}

func TestWinByPlacingAllMarkers(t *testing.T) {
	gd := loadOrSkip(t)
	s, _ := NewGame(Config{Data: gd, PlayerNames: []string{"A", "B"}, Seed: 9})
	total := s.Players[0].ControlMarkersTotal
	guard := 0
	for Winner(s) == "" {
		guard++
		if guard > 1000 {
			t.Fatal("game did not terminate")
		}
		// Whoever's turn it is places one marker.
		if err := RunTurn(s, gd, []Action{{Type: ActPlaceMarker}}); err != nil {
			t.Fatalf("RunTurn: %v", err)
		}
	}
	// Player A goes first and places one per turn, so A should win on turn `total`.
	if Winner(s) != "p1" {
		t.Errorf("winner = %q, want p1", Winner(s))
	}
	if s.Players[0].ControlMarkersPlaced != total {
		t.Errorf("winner placed %d markers, want %d", s.Players[0].ControlMarkersPlaced, total)
	}
}

func TestCannotOverplaceMarkers(t *testing.T) {
	gd := loadOrSkip(t)
	s, _ := NewGame(Config{Data: gd, PlayerNames: []string{"A", "B"}, Seed: 5})
	p := s.CurrentPlayerPtr()
	p.ControlMarkersPlaced = p.ControlMarkersTotal
	if err := Apply(s, gd, Action{Type: ActPlaceMarker}); err == nil {
		t.Errorf("expected error placing marker with none remaining")
	}
}

File: impl/ts/test/engine.test.ts
import { describe, expect, test } from "bun:test";
import { loadDefault } from "../src/data/index.ts";
import {
  controlMarkersFor,
  newGame,
  applyAction,
  advancePhase,
  applyActionWithEvents,
  runTurn,
  winner,
  type Config,
} from "../src/engine/index.ts";

const data = loadDefault("speedrunners");

function game(names: string[], seed: number) {
  const cfg: Config = { data, playerNames: names, seed };
  return newGame(cfg);
}

describe("control markers", () => {
  test("allotment by player count with default fallback", () => {
    expect(controlMarkersFor(data.mode, 2)).toBe(10);
    expect(controlMarkersFor(data.mode, 3)).toBe(8);
    expect(controlMarkersFor(data.mode, 4)).toBe(6);
    expect(controlMarkersFor(data.mode, 7)).toBe(8); // default
  });
});

describe("setup", () => {
  test("assigns players, distinct pawns, asymmetric opening deal", () => {
    const s = game(["Alice", "Bob"], 42);
    expect(s.players.length).toBe(2);
    for (const p of s.players) {
      expect(p.controlMarkersTotal).toBe(10);
      expect(p.pawnId).not.toBe("");
    }
    expect(s.players[0]!.pawnId).not.toBe(s.players[1]!.pawnId);
    // Asymmetric deal: p1 gets fewer than p2.
    expect(s.players[0]!.hand.length).toBeLessThan(s.players[1]!.hand.length);
  });
});

describe("determinism", () => {
  test("same seed -> same setup", () => {
    const a = game(["A", "B", "C"], 7);
    const b = game(["A", "B", "C"], 7);
    expect(a.players.map((p) => p.pawnId)).toEqual(b.players.map((p) => p.pawnId));
    expect(a.deck[0]).toBe(b.deck[0]!);
  });
});

describe("turn loop", () => {
  test("recycle refills hand to max", () => {
    const s = game(["A", "B"], 1);
    runTurn(s, data, [{ type: "pass" }]);
    expect(s.players[0]!.hand.length).toBe(s.players[0]!.maxHandSize);
  });

  test("turns advance and wrap", () => {
    const s = game(["A", "B", "C"], 3);
    expect(s.currentPlayer).toBe(0);
    runTurn(s, data);
    expect(s.currentPlayer).toBe(1);
    runTurn(s, data);
    runTurn(s, data);
    expect(s.currentPlayer).toBe(0);
  });
});

describe("live phase API", () => {
  test("advances through phases and reports recycle draws", () => {
    const s = game(["A", "B"], 41);
    expect(advancePhase(s, data).phase).toBe("action");
    const recycle = advancePhase(s, data);
    expect(recycle.phase).toBe("recycle");
    expect(recycle.events.some((event) => event.type === "draw")).toBe(true);
    expect(advancePhase(s, data).phase).toBe("end");
    expect(advancePhase(s, data).phase).toBe("beginning");
    expect(s.currentPlayer).toBe(1);
  });

  test("rejects illegal-phase and non-active-player actions without mutation", () => {
    const s = game(["A", "B"], 42);
    const before = s.players[0]!.controlMarkersPlaced;
    const blocked = applyActionWithEvents(s, data, { type: "place-marker" });
    expect(blocked.accepted).toBe(false);
    expect(blocked.events[0]!.type).toBe("validation-failed");
    expect(s.players[0]!.controlMarkersPlaced).toBe(before);
    advancePhase(s, data);
    const wrongPlayer = applyActionWithEvents(s, data, { type: "pass", playerId: "p2" });
    expect(wrongPlayer.error).toBe("only the active player may act");
  });

  test("reports accepted actions and state changes", () => {
    const s = game(["A", "B"], 43);
    advancePhase(s, data);
    const result = applyActionWithEvents(s, data, { type: "place-marker" });
    expect(result.accepted).toBe(true);
    expect(result.events.some((event) => event.type === "action-accepted")).toBe(true);
    expect(result.events.some((event) => event.type === "control-changed")).toBe(true);
  });
});

describe("win condition", () => {
  test("first to place all markers wins", () => {
    const s = game(["A", "B"], 9);
    const total = s.players[0]!.controlMarkersTotal;
    let guard = 0;
    while (!winner(s)) {
      if (++guard > 1000) throw new Error("game did not terminate");
      runTurn(s, data, [{ type: "place-marker" }]);
    }
    expect(winner(s)).toBe("p1");
    expect(s.players[0]!.controlMarkersPlaced).toBe(total);
  });

  test("cannot place more markers than available", () => {
    const s = game(["A", "B"], 5);
    const p = s.players[s.currentPlayer]!;
    p.controlMarkersPlaced = p.controlMarkersTotal;
    expect(() => applyAction(s, data, { type: "place-marker" })).toThrow();
  });
});
